### PR TITLE
PR for #65

### PR DESCRIPTION
**Pull request by max9901**
Hey guys, 

Here the PR to the development branch because i am not fully satisfied yet because there is no validation right now on the step-counting reset. 

However i think its a good first prototype. 

things i changed:
- add a step-count reset at midnight locale time
- store the last 7 days of steps in nvm memory and make them accessible trough the HAL object
- Make fetching the time for the RTC an stand alone process and add error checking here
- some cleanup to force even faster startups and some personal flavors probably. 

i would love to discuss the road forward if anyone is up-to it. 

cheers,
Max

### CODE REVIEW

File: src/main.cpp
Comment (Simonmicro) [inline src/main.cpp:150]: Is there any chance to finalize this before accepting this PR?

File: src/services/OswServiceTaskCore.cpp
Comment (Simonmicro) [inline src/services/OswServiceTaskCore.cpp:1]: Could your rename this task to `Steps`? As the task currently _only_ handles the step counting with the history and the name `core` is ambiguous.

File: include/osw_hal.h
Comment (Simonmicro) [inline include/osw_hal.h:8]: Could you try to use the Ardunio String library? It can be (always) used via `String` and does not require you to include the `std::string` lib.

File: src/apps/watchfaces/watchfacemax.cpp
Comment (Simonmicro) [inline src/apps/watchfaces/watchfacemax.cpp:5]: Same here.

**Review by Simonmicro — CHANGES_REQUESTED** [review]
Hey, nice work - here some requests, which must not be completely implemented, but would be very nice to have.

**Comment by Simonmicro** [comment]
Additionally could you try to rebase to the current `develop` branch?

Thx!

**Comment by uvwxy** [comment]
Nice work @max9901  , and thanks for the review @Simonmicro

File: src/main.cpp
Comment (max9901) [reply src/main.cpp:150]: Yeah that's easy but it's a trade-off we can just place it in the 30fps limiter of the screenflush. I wasn't sure which trade-off i wanted to make here. What would u suggest ?

File: src/services/OswServiceTaskCore.cpp
Comment (max9901) [reply src/services/OswServiceTaskCore.cpp:1]: Yeah my end goal is to actually perform all core functions here like:

```
  hal->handleWakeupFromLightSleep();
  hal->checkButtons();
  hal->fetchRtcTime();
 ```
What are your toughts ?

File: include/osw_hal.h
Comment (max9901) [reply include/osw_hal.h:8]: sure this wasn't me btw this is already in the development branch.

**Comment by max9901** [comment]
I already am in sync with the current `develop` branch right ? or am i missing something ?

File: include/osw_hal.h
Comment (Simonmicro) [reply include/osw_hal.h:8]: Okay that's on @rawsignal !

**Comment by Simonmicro** [comment]
You forked off a fork an therefore have an other history - which is fine for merging, but your history (and changes) would by much cleaner, when you only commit _your_ changes (meaning only after a60e180 ). That's something for the future, but also nice to correct on this one.

File: src/services/OswServiceTaskCore.cpp
Comment (Simonmicro) [reply src/services/OswServiceTaskCore.cpp:1]: >  hal->handleWakeupFromLightSleep();

YES

>  hal->fetchRtcTime();

Yes

>  hal->checkButtons();

Maybe? This is user interaction and therefore time critical - there is grantee that we won't slow down the services further down the road. You know what? I'll open an issue to rewrite the manager to respect time-critical services.

Comment (Simonmicro) [reply src/services/OswServiceTaskCore.cpp:1]: But for now, it does not has those - right? Therefore for now, it is only the step service. Also, why would you try to merge all those functions into one service task - wouldn't many small be better? Stack size... Yes... Okay, point made for now...

File: src/services/OswServiceTasks.cpp
Comment (Simonmicro) [inline src/services/OswServiceTasks.cpp:14]: Hey, this should stay - for anyone wanting to fiddle around with the example task (and learn how to use our services)!

File: src/main.cpp
Comment (Simonmicro) [reply src/main.cpp:150]: > This is btw the reason also why im have a PR to the develop branch :D

In that case, you could mark your PR as a draft. :grin: 

>  I wasn't sure which trade-off i wanted to make here.

Sry, but what are the tradeoffs? Fetch every cycle or... Every frame? The latter sounds better.

File: src/services/OswServiceTasks.cpp
Comment (max9901) [reply src/services/OswServiceTasks.cpp:14]: it's a few lines lower but it's still there.

File: src/main.cpp
Comment (max9901) [reply src/main.cpp:150]: if we include it in the display refresh things could break we certain apps don't update the screenbuffer as much. Simply because we fill the screen buffer first than than cal the display refresh function. We would be always a frame to late. and in the case of only 1 screenrefresh every 3 seconds for example this would break stuf.

Comment (Simonmicro) [reply src/main.cpp:150]: Ahh, got it. Okay - then it is fine there. Additionally the system should have always the newest time cached anyways.

File: src/services/OswServiceTasks.cpp
Comment (Simonmicro) [reply src/services/OswServiceTasks.cpp:14]: Whopps, missed that on my second review! :facepalm:

**Comment by max9901** [comment]
can we agree to merge this with a squash merge? so i don't need to rebase?

**Review by Simonmicro — APPROVED** [review]

**Comment by Simonmicro** [comment]
As this PR includes multiple (deep) changes (potentially breaking ones), I will not merge this on my own.

@uvwxy @SeanReg could you try to daily drive this for a while? I'll do this myself as soon it is in sync with the newest `develop` - @max9901 could you merge the current `develop` (and then ping me)? @uvwxy had fixed some stuff on there.

Some things to look out for:
* startup times really better (DeepS/LightS)?
* time still in sync (from wifi, manual, stopwatch)
* step-counter on zero when you pick it up
* something, something watchface?

**Comment by uvwxy** [comment]
@max9901 I would like to go over your changes and cherry pick them onto master. do you have some time this week to look over this together via discord/screenshare?

**Comment by max9901** [comment]
yes we can arrange that, can u send me a private message on discord ? so we can pick a timeslot. I can do tomorrow before 1500 the whole day and the rest of the week only in the evenings.

**Comment by uvwxy** [comment]
I think your user name differs on discord. You need to help me out a bit here 😅

**Comment by max9901** [comment]
maxe#2046   and im Max Eu in your channel.

**Comment by uvwxy** [comment]
Thank you a ton for your work!

I looked at your commit. squashing + merge / or merge / or rebase did not result in a mergeable result, I carried over your idea with db77c28 but running everything in the first core.

@Simonmicro rule of thumb: no i2c from background tasks to avoid collisions for now. all i2c reads should happen at the top of the main loop, or from an app. if we need data from an external device in the background we need to add semaphores.

**Pull request closed without merging**
